code/datums/components/aquarium.dm

```diff
@@ -0,0 +1,222 @@
+#define AQUARIUM_ANIMATION_FISH_SWIM "fish"
+#define AQUARIUM_ANIMATION_FISH_DEAD "dead"
+
+#define AQUARIUM_PROPERTIES_PX_MIN "px_min"
+#define AQUARIUM_PROPERTIES_PX_MAX "px_max"
+#define AQUARIUM_PROPERTIES_PY_MIN "py_min"
+#define AQUARIUM_PROPERTIES_PY_MAX "py_max"
+
+#define AQUARIUM_LAYER_MODE_BOTTOM "bottom"
+#define AQUARIUM_LAYER_MODE_TOP "top"
+#define AQUARIUM_LAYER_MODE_AUTO "auto"
+
+/datum/component/aquarium_content
+	/// This is a datum that describes our in-aquarium functionality
+	var/datum/aquarium_behaviour/properties
+
+	/// Keeps track of our current aquarium.
+	var/obj/structure/aquarium/current_aquarium
+
+	//This is visual effect holder that will end up in aquarium's vis_contents
+	var/obj/effect/vc_obj
+
+	/// Base px offset of the visual object in current aquarium aka current base position
+	var/base_px = 0
+	/// Base px offset of the visual object in current aquarium aka current base position
+	var/base_py = 0
+	//Current layer for the visual object
+	var/base_layer
+
+/datum/component/aquarium_content/Initialize(property_type)
+	if(!ismovable(parent))
+		return COMPONENT_INCOMPATIBLE
+	if(ispath(property_type,/datum/aquarium_behaviour))
+		properties = new property_type
+	else
+		CRASH("Invalid property type provided for aquarium content component")
+	properties.parent = src
+
+	RegisterSignal(parent, COMSIG_AQUARIUM_INSERT_READY, .proc/is_ready_to_insert)
+	RegisterSignal(parent, COMSIG_AQUARIUM_INSERTED, .proc/on_inserted)
+	RegisterSignal(parent, COMSIG_AQUARIUM_REMOVED, .proc/on_removed)
+	RegisterSignal(parent, COMSIG_AQUARIUM_FISH_CASE_STASIS, .proc/on_tank_stasis)
+
+/datum/component/aquarium_content/Destroy(force, silent)
+	if(current_aquarium)
+		remove_from_aquarium()
+	QDEL_NULL(vc_obj)
+	QDEL_NULL(properties)
+	return ..()
+
+/datum/component/aquarium_content/proc/is_ready_to_insert(datum/source,atom/aquarium_object)
+	SIGNAL_HANDLER
+	//This is kinda awful but we're unaware of other fish
+	if(properties.unique)
+		for(var/atom/movable/AM in aquarium_object)
+			if(AM == parent)
+				continue
+			var/datum/component/aquarium_content/other_content = AM.GetComponent(/datum/component/aquarium_content)
+			if(other_content && other_content.properties.type == properties.type)
```

Comment:
I'd rather have the aquarium keep track of whats in it rather than these getcomponents